lib/api/src/js/imports.rs

```diff
@@ -2,173 +2,207 @@
+    /// # fn foo_test(memory: Memory) {
+    /// let mut exports = Exports::new()
+    /// exports.insert("memory", memory);
     ///
-    /// import_object.register("namespace0", instance);
-    /// import_object.register("namespace1", namespace);
+    /// let mut import_object = Imports::new();
+    /// import_object.register_namespace("env", exports);
     /// // ...
+    /// # }
     /// ```
-    pub fn register<S, N>(&mut self, name: S, namespace: N) -> Option<Box<dyn LikeNamespace>>
-    where
-        S: Into<String>,
-        N: LikeNamespace + Send + Sync + 'static,
-    {
-        let mut guard = self.map.lock().unwrap();
-        let map = guard.borrow_mut();
-
-        match map.entry(name.into()) {
-            Entry::Vacant(empty) => {
-                empty.insert(Box::new(namespace));
-                None
-            }
-            Entry::Occupied(mut occupied) => Some(occupied.insert(Box::new(namespace))),
+    pub fn register_namespace(
+        &mut self,
+        ns: &str,
+        contents: impl IntoIterator<Item = (String, Extern)>,
+    ) {
+        for (name, extern_) in contents.into_iter() {
+            self.map.insert((ns.to_string(), name.clone()), extern_);
+        }
+    }
+
+    /// Add a single import with a namespace `ns` and name `name`.
+    ///
+    /// # Usage
+    /// ```no_run
+    /// # let store = Default::default();
+    /// use wasmer::{Imports, Function};
+    /// fn foo(n: i32) -> i32 {
+    ///     n
+    /// }
+    /// let mut import_object = Imports::new();
+    /// import_object.define("env", "foo", Function::new_native(foo));
+    /// ```
+    pub fn define(&mut self, ns: &str, name: &str, val: impl Into<Extern>) {
+        self.map
+            .insert((ns.to_string(), name.to_string()), val.into());
+    }
+
+    /// Returns the contents of a namespace as an `Exports`.
+    ///
+    /// Returns `None` if the namespace doesn't exist.
+    pub fn get_namespace_exports(&self, name: &str) -> Option<Exports> {
+        let ret: Exports = self
+            .map
+            .iter()
+            .filter(|((ns, _), _)| ns == name)
+            .map(|((_, name), e)| (name.clone(), e.clone()))
+            .collect();
+        if ret.is_empty() {
+            None
+        } else {
+            Some(ret)
         }
     }
 
-    fn get_objects(&self) -> VecDeque<((String, String), Export)> {
-        let mut out = VecDeque::new();
-        let guard = self.map.lock().unwrap();
-        let map = guard.borrow();
-        for (name, ns) in map.iter() {
-            for (id, exp) in ns.get_namespace_exports() {
-                out.push_back(((name.clone(), id), exp));
+    /// Resolve and return a vector of imports in the order they are defined in the `module`'s source code.
+    ///
+    /// This means the returned `Vec<Export>` might be a subset of the imports contained in `self`.
+    pub fn imports_for_module(&self, module: &Module) -> Result<Vec<Export>, InstantiationError> {
```

Comment:
We should be returning a `Vec<Extern>` here. Again, `Export` should be removed from the public API.